simSPI/linear_simulator/modules.py

```diff
@@ -0,0 +1,397 @@
+"""contains simulator and its components classes."""
+import os
+
+import mrcfile
+import numpy as np
+import torch
+import torch.fft
+from pytorch3d.transforms import Rotate
+
+from .utils import fourier_to_primal_2D, init_cube, primal_to_fourier_2D
+
+"""Module to generate data using using liner forward model."""
+
+
+class simulator(torch.nn.Module):
+    """Class to generate data using liner forward model.
+
+    Parameters
+    ----------
+    config: class
+        Class containing parameters of the simulator
+
+    """
+
+    def __init__(self, config, initial_volume=None):
+        super(simulator, self).__init__()
+
+        self.config = config
+        self.Projector = Projector(config)  # class for tomographic projector
+        self.init_volume()  # changes the volume inside the projector
+        self.CTF = CTF(config)  # class for ctf
+        self.Shift = Shift(config)  # class for shifts
+        self.Noise = Noise(config)  # class for noise
+
+    def forward(self, rot_params, ctf_params, shift_params):
+        """Create cryoEM measurements using input parameters.
+
+        Parameters
+        ----------
+        rot_params: dict of type str to {tensor}
+            Dictionary of rotation parameters for a projection chunk
+        ctf_params: dict of type str to {tensor}
+            Dictionary of Contrast Transfer Function (CTF) parameters
+             for a projection chunk
+        shift_params: dict of type str to {tensor}
+            Dictionary of shift parameters for a projection chunk
+
+        Returns
+        -------
+        projection.real : tensor
+            Tensor ([chunks,1,sidelen,sidelen]) contains cryoEM measurement
```

Comment:
is this the fastest way to do things vs [sidelen,sidelen,1,chunks]? What's the fast axis and slow axis?

Also, why the 4 dimensions, with the 1 in [chunks,1,sidelen,sidelen]?

Reply:
I'm not sure the order matters for speed (but I'm not an expert), though it does for broadcasting, which is also the reason why the second dimension exists here: https://pytorch.org/docs/stable/notes/broadcasting.html
To avoid confusion, I agree that it'd be nice to have a few words in the docstring on why this second dimension is needed for later.

Reply:
https://agilescientific.com/blog/2018/12/28/what-is-the-fastest-axis-of-an-array

Reply:
This is actually pytorch convention of keeping images in batch x channel x height x width format. This is important for example for the CNNs to understand which dimensions are batch, channel, etc. I have added a line in the docstring of this method in the new [PR](https://github.com/compSPI/simSPI/pull/23/).